### DISCUSSION TITLE
Use contracts instead of interfaces declarations in `Base_Test`

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/evm-monorepo#1062] [Ideas]
After the package tethering change, we’ve significantly increased the number of imports in `Base.t.sol`:

https://github.com/sablier-labs/v2-core/blob/a322ba684812fefb6fbf41cf6fd12e88904df662/test/Base.t.sol#L4-L32

$~$

One idea to reduce the number of imports (assuming we agree that having this many imports isn’t ideal) is to declare the testing contracts as contracts instead of interfaces (see the code below). Since the contracts themselves already inherit the interfaces, I don’t see any downside to this approach, as it would reduce the overall number of imports throughout the test files.

instead of:
```solidity
ISablierLockupDynamic internal lockupDynamic;
ISablierLockupLinear internal lockupLinear;
ISablierLockupTranched internal lockupTranched;
```

we could (option 1):
```solidity
SablierLockupDynamic internal lockupDynamic;
SablierLockupLinear internal lockupLinear;
SablierLockupTranched internal lockupTranched;
```

---

Another alternative to reducing the number of imports would be to import them directly from the contract (option 2):

```solidity
import { LockupNFTDescriptor, ILockupNFTDescriptor } from "src/core/LockupNFTDescriptor.sol";
import { SablierLockupDynamic, ISablierLockupDynamic } from "src/core/SablierLockupDynamic.sol";
import { SablierLockupLinear, ISablierLockupLinear } from "src/core/SablierLockupLinear.sol";
import { SablierLockupTranched, ISablierLockupTranched } from "src/core/SablierLockupTranched.sol";
```

Personally I chose option 1. wdyt @sablier-labs/solidity?

**Comment by smol-ninja** [comment]
Interesting proposal.

> Since the contracts themselves already inherit the interfaces, I don’t see any downside to this approach

By using interface as a datatype for a contract variable, it ensures that the public functions exposed by the contract are also included in the interface. This is because a contract can define more public function than the interface it inherits from. Consider the following case:

```solidity
interface IA {}

contract A is IA {
    function foo() public {}
}
```

If we only import `A`, the test would pass when it calls `foo` but if you import `IA` and define instance of `A` as `IA`, the test would fail when it calls `foo`. Thats the advantage option 1 would miss.

> assuming we agree that having this many imports isn’t ideal

I dont know what is ideal but having so many imports is common in testing: https://github.com/aave-dao/aave-v3-origin/blob/main/tests/utils/BatchTestProcedures.sol (our approach is rather cleaner than Aave's). I would not call it "not ideal" since its the base contract for testing so that should be fine imo.

**Reply by andreivladbrg** [reply]
> it ensures that the public functions exposed by the contract are also included in the interface

good point, i agree

> I would not call it "not ideal" since its the base contract for testing so that should be fine imo.

what do you say about option 2?

**Reply by smol-ninja** [reply]
Re option 2, no argument against it, just that I haven't seen an imports like these before where components are imported via inherited contracts. Its not typical to import through inherited components imo. But I suppose it works technically, right?

Also, chatgpt's input on this: https://chatgpt.com/share/670bdc9d-869c-8000-9368-7a22eea99fdd.

> Even though `LockupNFTDescriptor` inherits from `ILockupNFTDescriptor`, they are still distinct contracts/interfaces defined in separate files.

**Reply by andreivladbrg** [reply]
> But I suppose it works technically, right?

it does, yes

> > Even though `LockupNFTDescriptor` inherits from `ILockupNFTDescriptor`, they are still distinct contracts/interfaces defined in separate files.

good point.

---

i changed my mind about this proposal. it is better to keep the current imports closing
